### DISCUSSION TITLE
behavior of close

### DISCUSSION THREAD

**Discussion by junhyeok-yoon** [ANLAB-KAIST/KENSv3#151] [Project Q&A 2. Transport Layer (TCP)] [answered]
If I run solution and check packet log by wireshark,
We can observe that:
1. In backlog1, backlog2, client send FINACK and server send ACK and done
But In beforeAccept, afterAccept, in addition to that, server also send FINACK and client send ACK also.
I wonder why their behavior is different even though the test main code is same.
(In Korean)
backlog1, 2에서는 client가 더 이상 보내지 않겠다는 half close 상태로 끝나는 것 같고
beforeAccept, afterAccept에서는 client가 half close하고, server도 close해서 완전히 끊기는 것 같습니다.
왜 같은 server, client 코드를 쓰는데 close될 때 behavior가 다른지 궁금합니다.

2. It seems that the difference between before/afterAccept is the time when the server call accept.
Why this behavior leads to the different of the order of the packets?
(In Korean)
beforeAccept과 afterAccept의 차이가 accpet을 시도하는 시간 차이인것 같은데, 
packet log를 보면, before인 경우 (client)FINACK - (server)(ACK) - (server)(FINACK) - (client)(ACK)인데,
after는  (client)FINACK - (server)(FINACK) - (server)(ACK) - (client)(ACK)입닌다. 왜 이런 순서 차이가 나타는지 궁금합니다.
그리고 cloent는 finack을 보내고 server한테 ack을 받았으면 더 이상 안보낸다는건데 그 뒤의 ack은 어떻게 보낼 수 있는 것인가요? 더 이상 안보낸다는것이 data를 안보낸다는거지 control packet은 보낼 수 있는건가요?

3. In Tcp established state, there is no state transition when FINACK is received. Only if FIN received, it moves to CLOSE_WAIT. Please clarify this.

**Comment by Pusnow** [comment, marked as answer]
> In backlog1, backlog2, client send FINACK and server send ACK and done
But In beforeAccept, afterAccept, in addition to that, server also send FINACK and client send ACK also.
I wonder why their behavior is different even though the test main code is same.

This is related to the testing environment.
In backlog 1 and 2, the server is blocked in `accept()` and does not call `close()` until simulation ends.

Server expects 10:

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/app/kens/testhandshake.cpp#L205

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/app/kens/testhandshake.cpp#L278

But there are only 6 clients, and each connects to server only one:

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/app/kens/testhandshake.cpp#L247-L252

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/app/kens/testhandshake.cpp#L316-L321

Server is blocked here:

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/app/kens/testhandshake.cpp#L71

And server never calls `close()` until simulation ends:

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/app/kens/testhandshake.cpp#L102

So the server does not send FIN.



>It seems that the difference between before/afterAccept is the time when the server call accept.
Why this behavior leads to the different of the order of the packets?

In BeforeAccept, connection is established before accept, so the client can call `close()` even before the server calls `accept()`.

(client) call `close()` - (client)FINACK - (server)(ACK) - (server) call `accept()` - (server) call `close()` - (server)(FINACK) - (client)(ACK)

In AfterAccept, below happens.

1. (server) call `accept()`
2. (client) call `connect()`
3. (client) call `close()` / send FINACK
4. (server) receives ACK (of 3-way handshaking) / call `close()` / send FINACK
5. (server) receives FINACK / send ACK
6. (client) receives FINACK / send ACK



> 그리고 cloent는 finack을 보내고 server한테 ack을 받았으면 더 이상 안보낸다는건데 그 뒤의 ack은 어떻게 보낼 수 있는 것인가요? 더 이상 안보낸다는것이 data를 안보낸다는거지 control packet은 보낼 수 있는건가요?

Yes, in `FIN-WAIT-2` stage, client only can send control packets.




> In Tcp established state, there is no state transition when FINACK is received. Only if FIN received, it moves to CLOSE_WAIT. Please clarify this.

You can treat it as FIN.

**Reply by junhyeok-yoon** [reply]
I appreciate for your detail thorough answer!
